cypress/integration/core-opensearch-dashboards/opensearch-dashboards/apps/query_enhancements/saved_search.spec.js

```diff
@@ -0,0 +1,513 @@
+    ],
+    saveName: 'lucene-index-pattern-01',
+  },
+  'OpenSearch SQL': {
+    ...allowedSearchOperations['OpenSearch SQL'],
+    language: 'OpenSearch SQL',
+    dataset: INDEX_PATTERN_WITH_TIME,
+    queryString: `SELECT * FROM ${INDEX_PATTERN_WITH_TIME} WHERE bytes_transferred > 9950`,
+    hitCount: undefined,
+    sampleTableData: [],
+    saveName: 'sql-index-pattern-01',
+  },
+  PPL: {
+    ...allowedSearchOperations.PPL,
+    language: 'PPL',
+    dataset: INDEX_PATTERN_WITH_TIME,
+    queryString: `source = ${INDEX_PATTERN_WITH_TIME} | where bytes_transferred > 9950`,
+    hitCount: 101,
+    sampleTableData: [],
+    saveName: 'ppl-index-pattern-01',
+  },
+};
+const indexTestConfigurations = {
+  'OpenSearch SQL': {
+    ...allowedSearchOperations['OpenSearch SQL'],
+    language: 'OpenSearch SQL',
+    dataset: INDEX_WITH_TIME_1,
+    queryString: `SELECT * FROM ${INDEX_WITH_TIME_1} WHERE bytes_transferred > 9950`,
+    hitCount: undefined,
+    sampleTableData: [],
+    saveName: 'sql-index-01',
+  },
+  PPL: {
+    ...allowedSearchOperations.PPL,
+    language: 'PPL',
+    dataset: INDEX_WITH_TIME_1,
+    queryString: `source = ${INDEX_WITH_TIME_1} | where bytes_transferred > 9950`,
+    hitCount: 50,
+    sampleTableData: [],
+    saveName: 'ppl-index-01',
+  },
+};
+const allTestConfigurations = [
+  ...Object.values(indexPatternTestConfigurations),
+  ...Object.values(indexTestConfigurations),
+];
+
+// Maps a dataset that are used in this file to the exact string that dataset
+// corresponds to in the saved search API response
+const mapDatasetToType = (dataset) => {
+  switch (dataset) {
+    case INDEX_PATTERN_WITH_TIME:
+      return 'INDEX_PATTERN';
+    case INDEX_WITH_TIME_1:
+      return 'INDEXES';
+    default:
+      return 'unknown dataset';
+  }
+};
+
+// Maps a language that are used in this file to the exact string that the language
+// corresponds to in the saved search API response
+const mapLanguageToApiResponseString = (language) => {
+  switch (language) {
+    case 'DQL':
+      return 'kuery';
+    case 'Lucene':
+      return 'lucene';
+    case 'OpenSearch SQL':
+      return 'SQL';
+    case 'PPL':
+      return 'PPL';
+    default:
+      return 'unknown language';
+  }
+};
+
+// Escapes special characters for the editor
+const prepareQueryStringForEditor = (queryString) => {
+  return queryString.replaceAll(/([{}])/g, (char) => `{${char}}`);
+};
+
+// In theory this function is not needed as we should be able to use
+// the custom command navigateToWorkSpaceSpecificPage, but when using that
+// to go to the discover page, it sometimes leads to a blank page
+const navigateToDiscoverPage = () => {
+  cy.navigateToWorkSpaceHomePage(workspaceName);
+  cy.getElementByTestId('headerAppActionMenu').should('be.visible');
+};
+
+// Sets the dataset for the search. Since INDEXES are not saved to the dataset,
+// we need to click through various buttons to manually add them
+const setDataset = (dataset) => {
```

Comment:
Should we make mapLanguageToApiResponseString and mapDatasetToType to constants for type safe and single source of truth? For example, instead of running a func `mapDatasetToType` we just do 
```
const DatasetTypes = {
  INDEX_PATTERN: {
    name: 'INDEX_PATTERN',
    value: INDEX_PATTERN_WITH_TIME,
    supportedLanguages: [QueryLanguages.DQL, QueryLanguages.LUCENE, QueryLanguages.SQL, QueryLanguages.PPL],
  },
  INDEXES: {
    name: 'INDEXES',
    value: INDEX_WITH_TIME_1,
    supportedLanguages: [QueryLanguages.SQL, QueryLanguages.PPL],
  },
};
```
and 
```
const QueryLanguages = {
  DQL: {
    name: 'DQL',
    apiName: 'kuery',  // The mapping is part of the definition
    supports: {
      filters: true,
      histogram: true,
      // ...
    }
  },
  // ...
};
```
and `if (language === QueryLanguages.DQL.name)` instead of `if (language === 'DQL')`. So adding a new query language only requires updating QueryLanguages and relevant templates and
Adding a new dataset type is only add to DatasetTypes.

Then we can have 
TestCaseGenerator: Handles test case creation
TestExecutor: Manages test execution flow

```
// Test case generator for each dataset type and language combination
class TestCaseGenerator {
  static createQueryString(language, dataset, condition = 'bytes_transferred > 9950') {
    const queryTemplates = {
      [QueryLanguages.DQL.name]: condition,
      [QueryLanguages.LUCENE.name]: 'bytes_transferred: {9950 TO *}',
      [QueryLanguages.SQL.name]: `SELECT * FROM ${dataset} WHERE ${condition}`,
      [QueryLanguages.PPL.name]: `source = ${dataset} | where ${condition}`,
    };
    return queryTemplates[language];
  }

  static generateTestCase(dataset, language) {
    return {
      name: `${dataset.name}-${language.name}`,
      dataset: dataset.value,
      language: language.name,
      queryString: this.createQueryString(language.name, dataset.value),
      ...language.supports,
    };
  }

  static generateAllTestCases() {
    const testCases = [];
    
    Object.values(DatasetTypes).forEach(dataset => {
      dataset.supportedLanguages.forEach(language => {
        testCases.push(this.generateTestCase(dataset, language));
      });
    });
    
    return testCases;
  }
}

// Test execution helpers
class TestExecutor {
  static async setupTest(cy, testCase) {
    await this.navigateToDiscoverPage(cy);
    await this.setDataset(cy, testCase);
    await this.configureSearch(cy, testCase);
  }

  static async verifyTestResults(cy, testCase) {
    // Implementation of verification logic
    await this.verifyDiscoverPageState(cy, testCase);
    await this.verifySavedSearchInAssetsPage(cy, testCase);
  }
}
```
then 
```
// Main test suite
describe('Saved Search Tests', () => {
  beforeEach(() => {
    cy.setupTestData(SECONDARY_ENGINE.url, [...]);
    cy.addDataSource({
      name: TestSetup.datasourceName,
      url: SECONDARY_ENGINE.url,
      authType: 'no_auth',
    });
    cy.createInitialWorkspaceWithDataSource(TestSetup.datasourceName, TestSetup.workspaceName);
  });

  TestCaseGenerator.generateAllTestCases().forEach(testCase => {
    it(`should handle saved search for ${testCase.name}`, () => {
      TestExecutor.setupTest(cy, testCase);
      TestExecutor.verifyTestResults(cy, testCase);
    });
  });
});
```
I am thinking how to make the tests are readable and type safe and no big duplication.

Reply:
For:

```
const DatasetTypes = {
  INDEX_PATTERN: {
    name: 'INDEX_PATTERN',
    value: INDEX_PATTERN_WITH_TIME,
    supportedLanguages: [QueryLanguages.DQL, QueryLanguages.LUCENE, QueryLanguages.SQL, QueryLanguages.PPL],
  },
  INDEXES: {
    name: 'INDEXES',
    value: INDEX_WITH_TIME_1,
    supportedLanguages: [QueryLanguages.SQL, QueryLanguages.PPL],
  },
};
```

Im afraid of restricting the `value` to specific `INDEX_PATTERN_WQITH_TIME` and `INDEX_WITH_TIME_1`. For example, what if another dev wanted to use `INDEX_WITHOUT_TIME_1`? 

But I completely agree with abstracting things out that can be abstracted, and I will do so

Reply:
I implemented something very similar to what you have described. Could you take a look?